honeybadgermpc/protocols/avid.py

```diff
@@ -0,0 +1,225 @@
+# coding=utf-8
+import logging
+import math
+import asyncio
+from honeybadgermpc.protocols.reliablebroadcast \
+    import encode, decode, merkle_tree, get_merkle_branch, merkle_verify
+
+logger = logging.getLogger(__name__)
+logger.setLevel(logging.ERROR)
+# Uncomment this when you want logs from this file.
+# logger.setLevel(logging.NOTSET)
+
+
+class AVIDMessageType:
+    VAL = "VAL"
+    ECHO = "ECHO"
+    READY = "READY"
+    RETRIEVE = "RETRIEVE"
+    RESPONSE = "RESPONSE"
+
+
+class AVID:
+    # Class for Asynchronous Verifiable Information Dispersal
+    def __init__(self, n, t, leader, recv, send, input_size):
+        """ Initialize the class with parameters:
+        :param int n:  at least 3
+        :param int f: fault tolerance, ``N >= 3f + 1``
+        :param int leader: ``0 <= leader < N``
+        :param recv: :func:`recv()` blocks until a message is
+            recvd; message is of the form::
+            (i, (tag, ...)) = recv()
+            where ``tag`` is one of ``{"VAL", "ECHO", "READY"}``
+        :param send: sends (without blocking) a message to a designated
+            recipient ``send(i, (tag, ...))``
+        :param input_size: the size of the input messages to be retrieved
+        """
+        assert n >= 3 * t + 1
+        assert t >= 0
+        assert 0 <= leader < n
+
+        self._data = [None] * n
+        self.n = n
+        self.t = t
+        self.leader = leader
+        self.recv = recv
+        self.send = send
+        # size of input_list for disperse
+        self.input_size = input_size
+        # the response for retrieve
+        self.retrieval_queue = asyncio.Queue()
+        # A Queue for OK: retrive will be sent until the ok state is reached
+        self.ok_queue = asyncio.Queue()
+
+    def broadcast(self, o):
+        for i in range(self.n):
+            self.send(i, o)
+
+    async def retrieve(self, sid, index):
+        """ Retrieve individual item
+        :param int index:  the index of retrieval
+        """
+        # we are able to retrieve things from AVID indivdually
+        assert 0 <= index < self.input_size
+
+        # send retrieve until ok state
+        assert await self.ok_queue.get()
+
+        # send retrieve to all the parties
+        self.broadcast((sid, AVIDMessageType.RETRIEVE, index))
+
+        result = [None] * self.input_size
+        response_set = set()
+
+        # response threshold same as k
+        response_threshold = self.n - 2 * self.t
+
+        def decode_output(roothash):
+            decoded_output = decode(response_threshold, self.n, result)
+            # Rebuild the merkle tree to guarantee decoding is correct
+            _stripes = encode(response_threshold, self.n, decoded_output)
+            _mt = merkle_tree(_stripes)
+            _roothash = _mt[1]
+            try:
+                assert _roothash == roothash
```

Comment:
Again, no `assert` when you want to validate something. Asserts are only meant for testing and are disabled when running the code. you should do 
```
if _roothash == roothash:
    raise
```